### PR TITLE
document the implications of using the `insecure` overlap strategy for cockroachdb

### PR DESCRIPTION
**Pull request by ecordell**

### CODE REVIEW

File: internal/datastore/crdb/README.md [hunk 1]
```diff
@@ -8,3 +8,103 @@ The CockroachDB implementation should be used when your SpiceDB service runs in
...
+
+Same as before, the normal case is that the timestamp for `A` < the timestamp for `B`, but if the two writes hit the conditions for a timestamp reversal, then `B < A`.
+
+Bob later tries to read the document. The application performs an `AtLeastAsFresh` `Check` for Bob to access the document `secret` using the stored Zedtoken (which is timestamp `B`.)
+
+It's possible that SpiceDB will pick an evaluation timestamp `T` such that `B < T < A`, so that bob is allowed to read the newest contents of the document, and discover that Alice thinks he is a fool.
+
+Same as before, this is only possible if `A - T < quantization window`: Bob's check has to happen soon enough after the write for `A` that it's possible that SpiceDB picks a timestamp in between `A` and `B`, and the default quantization window is `5s`.
+
+### Application Mitigations for Misapplying Old ACLs
```
Comment (vroldanbet) [inline internal/datastore/crdb/README.md:105]: nit: may be better to make this a subsection of `Mis-apply Old ACLs to New Content`

```suggestion
#### Application Mitigations for Misapplying Old ACLs
```

**Review by vroldanbet — COMMENTED** [review]
LGTM, left some nits and questions. I think I'd be worth adding this also to https://authzed.com/docs/spicedb/selecting-a-datastore#cockroachdb

File: internal/datastore/crdb/README.md [hunk 2]
```diff
@@ -8,3 +8,103 @@ The CockroachDB implementation should be used when your SpiceDB service runs in
...
+- And those two writes hit two different gateway nodes
+- And the CRDB gateway node clocks have a delta `D`
+- And the writes touch disjoint sets of relationships
+- And those two writes are sent within the time delta `D` between the gateway nodes
+- And the writes land in ranges whose followers are disjoint sets of nodes
+- And other independent cockroach processes (heartbeats, etc) haven't coincidentally synced the gateway node clocks during the writes.
+
+Then it's possible that the second write will be assigned a timestamp earlier than the first write. In the next section we'll look at whether that matters for your application, but for now let's look at what makes the above conditions more or less likely:
+
+- **Clock skew**. A larger clock skew gives a bigger window in which timestamps can be reversed. But note that CRDB enforces a max offset between clocks, and getting within some fraction of that max offset will boot the node from the cluster.
```
Comment (josephschorr) [inline internal/datastore/crdb/README.md:43]: "boot" -> "kick"

File: internal/datastore/crdb/README.md [hunk 3]
```diff
@@ -8,3 +8,103 @@ The CockroachDB implementation should be used when your SpiceDB service runs in
...
+Let's look at how likely this is, and what the impact might actually be for your workload.
+
+## When can timestamps be reversed?
+
+Before we look at how this can impact an application, let's first understand when and how timestamps can be reversed in the first place.
+
+- When two writes are made in short succession against CockroachDB
+- And those two writes hit two different gateway nodes
+- And the CRDB gateway node clocks have a delta `D`
+- And the writes touch disjoint sets of relationships
```
[the remark this answers is not in the document]
Comment (ecordell) [reply internal/datastore/crdb/README.md:36]: Nope, if the the writes land in the same ranges (which they would if any relationships overlap) then the timestamps will get synchronized

**Review by jakedt — CHANGES_REQUESTED** [review]
You might also want to address the fact that timestamp reversals on disjoint graphs (not just non-overlapping) will never result in the new enemy problem. Maybe write up some examples on when there wouldn't be any impact whatsoever.

**Comment by ecordell** [comment]
@jakedt updated with examples like you suggested

**Review by jakedt — APPROVED** [review]

**Comment by ecordell** [comment]
@jakedt fixed

**Review by jakedt — APPROVED** [review]

**Pull request merged**